### DISCUSSION TITLE
MappedQueue in greedy_modularity_communities()

### DISCUSSION THREAD

**Discussion by elplatt** [networkx/networkx#5083] [General]
This is a continuation of the discussion in #4992, #5000, and #5065. CC @dschult @tristanic @ThanasisMattas 

PR #5065 makes a few changes in order to fix #4992:
1. It avoids assuming that `-(-dq) == dq`, which was failing due to a rounding error, by treating `dq` and `negdq` as separate.
2. It changes `MappedQueue` to allow separate tracking of priority and data.

It appears that the first change is all that is necessary to fix the error. The second change doesn't seem to have any advantage in this case and comes with several disadvantages:
* Additional functional calls which are expensive in python.
* Adds complexity to the `MappedQueue` class.
* The added `_HeapElement` class allows weird behavior where `a >= b and a <= b` is true but `a == b` is false.

My main question is wether separating the priority and the data is really necessary here. I understand why it might be desirable in some applications, but I don't see the benefit in this particular case. I understand the original reasoning was the concern that floating point keys could cause problems because of rounding. However, rounding only occurs when arithmetic is performed, and keys should be treated as immutable (floating point or otherwise), so rounding should never occur. As I understand it, the cause of the bug was the multiplication by `-1` to convert `-dq` back to `dq` (which was my fault, sorry!) and that has been fixed independently.

Ideally, I would like to keep the first change, but restore the original `MappedQueue` code. But if there is a compelling reason not to, I'm open to that. In that case I would still like to restore `MappedQueue` to the original version, and encapsulate the separation of data and priority entirely in `_HeapElement`, which will make tracking down future bugs easier and avoid the additional overhead when `MappedQueue` data and priority are the same.

So is there really a need to separate data and priority. If so, what reasons am I not seeing?

**Comment by dschult** [comment]
Thanks for this post -- I'd like to investigate this more if possible -- there may also be speed-ups that you mentioned in a different post.

Why do you say that the only arithmetic used in this algorithm is setting `dq` to `-dq`?  Also, how would renaming `dq` to `negdq` help with that? Aren't we still negating `dq` twice (whatever the variable is called)?  I may be missing something.

The next question I have is how we know that the shifts and updates to the weights that occur in the algorithm don't also cause round-off error? e.g. the resulting `dq_vw` value might be supposed to match a value from the queue, but doesn't due to round-off.

As far as `_HeapElement` having strange behavior, when I think about it from the right perspective, I think that the confusing inequalities are due to wanting both comparison based on priority AND checking equality based on node-pairs. Making it a single object instead of two is the cause of the confusion. Perhaps there is a better (and probably faster) way to finesse the separation of the two. I suspect a brute force separation (like with a dict keyed by node-pairs to the priority) could work for this.

Also, using 3-tuples in comparisons during a heap sift requires the graph's nodes be ordered. And that is not always true. For example if the Graph has strings and integers as nodes...  Or even Graphs as nodes in a super-graph. This was a subtle bug in the original MappedQueue. I think it could be finessed by making a 4-tuple with unique integer for each node-pair or something...    

But it might be best to move to something like `heapdict`.  Let's follow up on this and get the best solution -- it probably isn't what we've put in place as a fix for 2.6.3.
